python/nx-cugraph/nx_cugraph/algorithms/cluster.py

```diff
@@ -0,0 +1,139 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import cupy as cp
+import pylibcugraph as plc
+
+from nx_cugraph.convert import _to_undirected_graph
+from nx_cugraph.utils import networkx_algorithm, not_implemented_for
+
+__all__ = [
+    "triangles",
+    "average_clustering",
+    "clustering",
+    "transitivity",
+]
+
+
+def _triangles(G, nodes, symmetrize=None):
+    if nodes is not None:
+        if is_single_node := (nodes in G):
+            nodes = [nodes if G.key_to_id is None else G.key_to_id[nodes]]
+        else:
+            nodes = list(nodes)
+        nodes = G._list_to_nodearray(nodes)
+    else:
+        is_single_node = False
+    if len(G) == 0:
+        return None, None, is_single_node
+    node_ids, triangles = plc.triangle_count(
+        resource_handle=plc.ResourceHandle(),
+        graph=G._get_plc_graph(symmetrize=symmetrize),
+        start_list=nodes,
+        do_expensive_check=False,
+    )
+    return node_ids, triangles, is_single_node
+
+
+@not_implemented_for("directed")
+@networkx_algorithm(plc="triangle_count", version_added="24.02")
+def triangles(G, nodes=None):
+    G = _to_undirected_graph(G)
+    node_ids, triangles, is_single_node = _triangles(G, nodes)
+    if len(G) == 0:
+        return {}
+    if is_single_node:
+        return int(triangles[0])
+    return G._nodearrays_to_dict(node_ids, triangles)
+
+
+@not_implemented_for("directed")
+@networkx_algorithm(is_incomplete=True, plc="triangle_count", version_added="24.02")
+def clustering(G, nodes=None, weight=None):
+    """Directed graphs and `weight` parameter are not yet supported."""
+    G = _to_undirected_graph(G)
+    node_ids, triangles, is_single_node = _triangles(G, nodes)
+    if len(G) == 0:
+        return {}
+    if is_single_node:
+        numer = int(triangles[0])
+        if numer == 0:
+            return 0
+        degree = int((G.src_indices == nodes).sum())
+        return 2 * numer / (degree * (degree - 1))
+    # What about self-edges?
```

Comment:
This leaves me curious; what happens with self-edges? I'm guessing the degree value for a node with a self-edge might be artificially high, but I hate to guess. Is this a limitation we need to address (perhaps as a FIXME) and/or something we need to include in `extra_docstrings`? Maybe at a minimum can you update the comment to elaborate?

Reply:
NetworkX is underspecified. There are no tests with self-edges. Perhaps we could add a test to match NetworkX, _or_ we could try to upstream tests to NetworkX. The latter will require the NetworkX team being intentional about what is supposed to happen in this case. Should we ignore self-edges or not?

Reply:
@eriknw and I discussed offline and both think self-edges probably should not be included.

However, there's a discussion about this started [here](https://github.com/networkx/networkx/discussions/7238) which could change this (probably in a future PR if not answered soon).

Reply:
updated to ignore selfloops and comparison tests added.